packages/mui-joy/src/IconButton/IconButton.tsx

```diff
@@ -0,0 +1,197 @@
+import { styled, useThemeProps } from '../styles';
+import iconButtonClasses, { getIconButtonUtilityClass } from './iconButtonClasses';
+import { IconButtonProps, IconButtonTypeMap, ExtendIconButton } from './IconButtonProps';
+
+const useUtilityClasses = (ownerState: IconButtonProps & { focusVisible: boolean }) => {
+  const { color, disabled, focusVisible, focusVisibleClassName, size, shape, variant } = ownerState;
+
+  const slots = {
+    root: [
+      'root',
+      disabled && 'disabled',
+      focusVisible && 'focusVisible',
+      variant && `variant${capitalize(variant)}`,
+      color && `color${capitalize(color)}`,
+      size && `size${capitalize(size)}`,
+      shape && `shape${capitalize(shape)}`,
+    ],
+  };
+
+  const composedClasses = composeClasses(slots, getIconButtonUtilityClass, {});
+
+  if (focusVisible && focusVisibleClassName) {
+    composedClasses.root += ` ${focusVisibleClassName}`;
+  }
+
+  return composedClasses;
+};
+
+const IconButtonRoot = styled('button', {
+  name: 'MuiIconButton',
+  slot: 'Root',
+  overridesResolver: (props, styles) => styles.root,
+})<{ ownerState: IconButtonProps }>(({ theme, ownerState }) => [
+  {
+    '--IconButton-size': '2.5rem',
+    padding: '0.25rem',
+    ...(ownerState.variant === 'outlined' && {
+      padding: 'calc(0.25rem - 1px)', // account for the border width
+    }),
+    ...(ownerState.size === 'sm' && {
+      '--IconButton-size': '2rem',
+      // same padding as size `md`
+    }),
+    ...(ownerState.size === 'lg' && {
+      '--IconButton-size': '3rem',
+      padding: '0.5rem',
+      ...(ownerState.variant === 'outlined' && {
+        padding: 'calc(0.5rem - 1px)', // account for the border width
+      }),
+    }),
+    minWidth: 'var(--IconButton-size)', // use min-width instead of height to make the button resilient to its content
+    minHeight: 'var(--IconButton-size)', // use min-height instead of height to make the button resilient to its content
+    borderRadius: theme.vars.radius.sm,
+    ...(ownerState.shape === 'square' && {
+      borderRadius: 0,
+    }),
+    ...(ownerState.shape === 'circular' && {
+      borderRadius: 'var(--IconButton-size)', // don't divided by 2, in case the icon causes the size to be bigger so that the button remains circular
+    }),
+    border: 'none',
+    backgroundColor: 'transparent',
+    display: 'inline-flex',
+    alignItems: 'center',
+    justifyContent: 'center',
+    position: 'relative',
+    // TODO: discuss the transition approach in a separate PR. This value is copied from mui-material Button.
+    transition:
+      'background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms, box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms, border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms, color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
+    [`&.${iconButtonClasses.focusVisible}`]: theme.focus.default,
+  },
+  theme.variants[ownerState.variant!]?.[ownerState.color!],
```

Comment:
Hmm, what about if developers add their own theme? We should document that they need to override the `theme.variants` too. This will be tricky if they want to add variant for only one component.

Reply:
> what about if developers add their own theme?

Yes, I assume that you mean another variant.
- They can do at the global level (theme) and apply to all components that support variant prop
  ```js
  <CssVarsProvider theme={{
    variants: {
      foo: {
        primary: { ...},
        ...other color,
      },
      fooHover: {
        primary: {...},
        ...other color,
      }
    }
  }}>
  ```
- At the component level (augment the variant types). The nice part of this is that the component styles won't blow up (preserve layout, size and display) because our variant only handle the colors
  ```js
  <CssVarsProvider theme={{
    components: {
      MuiButton: {
        styleOverrides: {
          root: ({ ownerState }) => ({
            ...ownerState.variant === 'foo' && {
              // custom styles for variant foo
            }
          })
        }
      }
    }
  }}>
  ```